File: 05-1.rb
#!/usr/bin/env ruby

class Map
  attr_accessor :from, :to, :map

  def initialize(from, to)
    @from = from
    @to = to
    @map = []
  end

  def add_mapping(source, destination, length)
    @map << {source: source, destination: destination, length: length}
  end

  def covers?(source, length, test)
    test >= source and test < source + length
  end

  def [](input)
    @map.each do |mapping|
      if covers?(mapping[:source], mapping[:length], input)
        return mapping[:destination] - mapping[:source] + input
      end
    end
    input
  end

  def to_s
    s = "<#{self.class} #{@from}-#{to}: "
    @map.each do |m|
      s += "#{m[:source]} #{m[:length]} #{m[:destination]} "
    end
    s += ">"
    s
  end

end

input = File.read('05.input').lines.map(&:strip)

seeds = nil
maps = []
map = nil

input.each do |line|
  if line.start_with? 'seeds: '
    seeds = line.split(':')[1].split.map(&:to_i)
  elsif line.empty?
    maps << map if map
    map = nil
  elsif line.end_with? 'map:'
    names = line.split[0].split('-')
    map = Map.new(names[0], names[2])
  else
    mapping = line.split.map(&:to_i)
    map.add_mapping(mapping[1], mapping[0], mapping[2])
  end
end
maps << map

locations = seeds.map do |seed|
  from = 'seed'
  pos = seed
  loop do
    map = maps.find { |m| m.from == from }
    to = map.to
    pos = map[pos]
    break if to == 'location'
    from = to
  end
  pos
end

print locations.min, "\n"

File: 05-1.py
#!/usr/bin/env python

class Map:
    def __init__(self, from_type, to_type):
        self.from_type = from_type
        self.to_type = to_type
        self.mappings = []

    def add_mapping(self, source, destination, length):
        self.mappings.append({'source': source, 'destination': destination, 'length': length})

    def covers(self, source, length, test):
        return test >= source and \
            test < source + length

    def __getitem__(self, input):
        for mapping in self.mappings:
            if self.covers(mapping['source'], mapping['length'], input):
                return mapping['destination'] - mapping['source'] + input
        return input

    def __str__(self):
        s = f"<{self.__class__.__name__} {self.from_type}-{to_type}: "
        for m in self.mappings:
            s += f"{m['source']} {m['length']} {m['destination']} "
        s += ">"
        return s


input = [row.strip() for row in open('05.input').readlines()]

seeds = None
maps = []
mapping = None

for line in input:
    if line.startswith('seeds: '):
        seeds = [int(s) for s in line.split(':')[1].split()]
    elif len(line) == 0:
        if mapping is not None:
            maps.append(mapping)
        mapping = None
    elif line.endswith('map:'):
        names = line.split()[0].split('-')
        mapping = Map(names[0], names[2])
    else:
        nums = [int(n) for n in line.split()]
        mapping.add_mapping(nums[1], nums[0], nums[2])

maps.append(mapping)

locations = []
for seed in seeds:
    from_type = 'seed'
    pos = seed
    while True:
        mapping = None
        for m in maps:
            if m.from_type == from_type:
                mapping = m
                break
        to_type = mapping.to_type
        pos = mapping[pos]
        if to_type == 'location':
            break
        from_type = to_type
    locations.append(pos)

print(min(locations))
